src/Raven.Server/Web/RequestHandler.cs

```diff
@@ -687,6 +689,61 @@ private static void ThrowInvalidAuthStatus(RavenServer.AuthenticationStatus? sta
             throw new ArgumentOutOfRangeException("Unknown authentication status: " + status);
         }
 
+        public static bool CheckCSRF(HttpContext httpContext, ServerStore serverStore)
+        {
+            if (serverStore.Configuration.Security.EnableCsrfFilter == false)
+                return true;
+            
+            var host = httpContext.Request.Host;
+            if (string.IsNullOrWhiteSpace(host.ToString()))
+                return false;
+            
+            var requestedOrigin = httpContext.Request.Headers["Origin"];
+            
+            if (requestedOrigin.Count == 0 || requestedOrigin[0] == null)
+                return true;
+
+            if (SafeCsrfMethods.Contains(httpContext.Request.Method))
+                return true;
+            
+            var origin = requestedOrigin[0];
+            var originHost = new Uri(origin).Host;
+            var originAuthority = new Uri(origin).Authority;
```

Comment:
double URI parsing here, should we cache it?

Reply:
done